ownership/sync-repos.py

```diff
@@ -0,0 +1,255 @@
+    # moved or archived
+    for ss_name in names_only_ss.copy():
+        try:
+            found_repo = hub.repository(*ss_name.split('/', 2))
+        except NotFoundError:
+            found_repo = None
+        if not found_repo:
+            recommend_delete_row(ss_name, "In spreadsheet but not Github, and not moved")
+        elif str(found_repo.owner) not in orgs:
+            recommend_delete_row(
+                ss_name, "Moved to a Github org not in the scan list",
+                {'new_location': str(found_repo)})
+        elif found_repo.archived:
+            recommend_delete_row(ss_name, "Repo has been archived")
+        else:
+            new_name = str(found_repo).lower()
+            if [r for r in repos_ss if r['name'] == new_name]:
+                bail("Manual intervention required: A repo that has moved "
+                     "exists in the spreadsheet under both its old name "
+                     "and its new name. %s -> %s"
+                     % (ss_name, new_name))
+            actions.append({
+                'action': 'rename_row',
+                'name': ss_name,
+                'new_name': new_name,
+                'row_id': find_row_by_name(ss_name)['row_id'],
+                'why': ("Repo has been renamed, and is not already in the "
+                        "spreadsheet under that name")
+            })
+            names_only_ss.remove(ss_name)
+            names_only_gh.remove(new_name)
+    
+    # Now check ones that are only in Github. We've already filtered
+    # out archived repos and removed any that are pending a rename on
+    # the spreadsheet.
+    for gh_name in names_only_gh.copy():
+        actions.append({
+            'action': 'add_row',
+            'name': gh_name,
+            'why': "Repo present in Github but not spreadsheet"
+        })
+        names_only_gh.remove(gh_name)
+
+    # Make sure nothing is unaccounted for before presenting recommendations.
+    if names_only_gh or names_only_ss:
+        bail("Failed to convert all mismatches between Github and spreadsheet "
+             "into actions! Github remaining = %s; spreadsheet remaining = %s"
+             % (names_only_gh, names_only_ss))
+
+    # Ensure actions don't have any internal conflicts. It might be OK
+    # if the same *name* comes up twice; for example, we might want to
+    # move an archived repo to the unspported org *and* remove it from
+    # the spreadsheet. However, two actions for the same row ID is a
+    # problem.
+    seen_row_ids = set()
+    for action in actions:
+        row_id = action.get('row_id')
+        if not row_id:
+            continue
+        if row_id in seen_row_ids:
+            bail("Possible bug in script: Was going to suggest multiple "
+                 "actions for row %s: %s"
+                 % (row_id, [r for r in actions if r.get('row_id') == row_id]))
+        seen_row_ids.add(row_id)
+
+    # For now, just make recommendations, don't make changes.
+    if actions:
```

Comment:
Should this be split out from the run-compare?  It seems like there is a nice seam here where the compare outputs the diff + actions and than what we do with them can be de-coupled? Might be a premature optimization though so fine to leave it as is too.

Reply:
Agreed, this really shouldn't be in `run_compare`, but rather `actions` should be returned to `main`, and then passed to another function that actually makes changes. Easy enough to make the split right now, though.